workspace-manager-clienttests/src/main/java/scripts/utils/WorkspaceManagerServiceUtils.java

```diff
@@ -0,0 +1,106 @@
+                server.testRunnerServiceAccount.name);
+
+        // build the client object
+        ApiClient apiClient = new ApiClient();
+        apiClient.setBasePath(server.workspaceManagerUri);
+        apiClient.setAccessToken(accessToken.getTokenValue());
+
+        return apiClient;
+    }
+
+    /**
+     * Build the Workspace Manager API client object for the given test user and server specifications. This
+     * class maintains a cache of API client objects, and will return the cached object if it already
+     * exists. The token is always refreshed, regardless of whether the API client object was found in
+     * the cache or not.
+     *
+     * @param testUser the test user whose credentials are supplied to the API client object
+     * @param server the server we are testing against
+     * @return the API client object for this user
+     */
+    public static ApiClient getClientForTestUser(
+            TestUserSpecification testUser, ServerSpecification server) throws IOException {
+        if (server.workspaceManagerUri == null || server.workspaceManagerUri.isEmpty()) {
+            throw new IllegalArgumentException("Workspace Manager URI cannot be empty");
+        }
+
+        // if no test user is specified, then return a client object without an access token set
+        // this is useful if the caller wants to make ONLY unauthenticated calls
+        if (testUser == null) {
+            ApiClient apiClient = new ApiClient();
+            apiClient.setBasePath(server.workspaceManagerUri);
+        }
+
+        // refresh the user token
+        GoogleCredentials userCredential =
+                AuthenticationUtils.getDelegatedUserCredential(
+                        testUser, AuthenticationUtils.userLoginScopes);
+        AccessToken userAccessToken = AuthenticationUtils.getAccessToken(userCredential);
+        // first check if there is already a cached ApiClient for this test user
+        ApiClient cachedApiClient = apiClientsForTestUsers.get(testUser);
+        if (cachedApiClient != null) {
+            // refresh the token here before returning
+            // this should be helpful for long-running tests (roughly > 1hr)
+            cachedApiClient.setAccessToken(userAccessToken.getTokenValue());
+
+            return cachedApiClient;
+        }
+
+        // TODO: have ApiClients share an HTTP client, or one per each is ok?
+        // no cached ApiClient found, so build a new one here and add it to the cache before returning
+        logger.debug(
+                "Fetching credentials and building Workspace Manager ApiClient object for test user: {}",
+                testUser.name);
+        ApiClient apiClient = new ApiClient();
+        apiClient.setBasePath(server.workspaceManagerUri);
+
+        apiClient.setAccessToken(userAccessToken.getTokenValue());
+
+        apiClientsForTestUsers.put(testUser, apiClient);
+        return apiClient;
+    }
```

Comment:
it feels dangerous to reuse ApiClients in a parallel/multithreaded test. Is there a good reason to cache clients and reuse them? I don't think ApiClients are threadsafe or reusable.

Consider `ServiceStatus` line 27:
```
int httpCode = unauthenticatedApi.getApiClient().getStatusCode();
```
when you invoke a request via the api client, it saves the http response code to a member variable of the ApiClient class. If you are reusing ApiClients across multiple requests, those requests can overwrite each other.

If I am wrong about this, please show me how!

Reply:
After discussion with Mariko, we simply removed caching part from the code.